blog/2024-03-15-structuring-a-large-app.md

```diff
@@ -0,0 +1,370 @@
+---
+author: Masen Furer
+date: 2024-03-15
+title: "Structuring a Large App"
+description: "So your Reflex app is getting large? Here's some advice on how to lay it out."
+image: /blog/alek99_squares_abstract_animated._dark_purple_with_yellow_highl_976e3d65-b044-48e6-880e-b7983b5e0b42.webp
+---
+
+## First Steps
+
+Reflex makes it easy to create a simple Web UI running from a single Python file:
+
+```python
+import reflex as rx
+
+
+@rx.page()
+def index():
+    return rx.center(
+        rx.heading("Hello Reflex! 👋")
+    )
+
+
+app = rx.App()
+```
+
+However, as the app grows, keeping everything in a single module can quickly
+become messy and unmanageable.
+
+Because a Reflex app is just a Python project, it can be organized into modules
+and packages (directory with `__init__.py`) and imported like any other Python objects.
+
+## What Makes a Reflex App?
+
+When executing `reflex init` or `reflex run`, the framework will import **two** modules automatically.
+
+### rxconfig.py
+
+Reflex first looks for an `rxconfig.py` in the current working directory and
+imports it.  At minimum, it **must define a module-level global named `config` as
+an instance of `rx.Config`**, of which the only required setting is `app_name`.
+
+```python
+import reflex as rx
+
+config = rx.Config(app_name="example_big_app")
+```
+
+### \{app_name}/\{app_name}.py
+
+After loading the config, Reflex imports the main app module based on the
+`app_name`, which **must define a module-level global named `app` as an instance
+of `rx.App`**.
+
+If the app_name is `example_big_app`, then the main module would be 
+`example_big_app/example_big_app.py`.
+
+The main app module is responsible for importing all other modules that make up the
+app and defining `app = rx.App()`. **All other modules containing pages, state,
+and models MUST be imported by the main app module or package** for Reflex to include
+them in the compiled output.
+
+## Breaking the App into Modules
+
+The key to good organization as the app grows is to break down all parts of the app into
+smaller, more manageable pieces, and organize them into logical packages that avoid circular
+dependencies.
+
+### Pages
+
+#### example_big_app/template.py
```

Comment:
Maybe before introducing templates, show factoring out pages into multiple files. Then we can note that the pages have something in common which leads to the template. (pretty much reordering this with the next section, since multiple pages is the more basic use case than templating)